keps/sig-node/20190226-migrate-in-tree-dockershim-to-out-of-tree.md

```diff
@@ -0,0 +1,148 @@
+- The number of affected users maybe large.
+- Updating all the eco-system tools to support the new cri-dockerd.
+- Many people use the built in dockershim for in-cluster image build. While that may not be something we recommend for a variety of reasons, it will be a breaking change for these users.
+- CRI is still in alpha，should probably get a 1.0 out there splitting out dockershim completely from kubelet.
+- Existed CNI and CSI plugins may also be affected.
+
+> You can check [the discussion in sig-node mailing list](https://groups.google.com/forum/#!msg/kubernetes-sig-node/0qVzfugYhro/l6Au216XAgAJ) for more details. 
+
+Based on all the discussion, we agree that we should not rush to immediate decision. At the same time, it's the right time to start designing and documenting dockershim deprecation criteria and plan, which will be the main content of rest of this KEP.
+
+### Goals
+
+- A concrete dockershim deprecation criteria.
+- A brief plan to deprecate dockershim spanning multiple releases.
+- Issues need to be fixed in kubelet in order to separate dockershim from it.
+
+### Non-Goals
+
+- Deprecation of dockershim immediately without consideration for users and WIP efforts depending on it.
+- Refactoring or re-design of dockershim itself due to deprecation.
+
+## Proposal
+
+### Dockershim deprecation criteria
+
+- CRI itself is beta.
+- kubelet has no dependency on dockershim/docker in its whole lifecycle. 
+- All node related features are CRI generic and have no "back door" dependency on dockershim/docker.
+- Reasonable benchmark result of performance degradation after moving dockershim to out-of-tree.
+- A out-of-tree CRI for docker is implemented and well maintained, and become to beta.
+- E2E test framework has been updated with fully support of out-of-tree CRI container runtime.
+
+### Dockershim deprecation plan
+
+Step 1: Stabilize in-tree dockershim and decouple dockershim from kubelet (but still in-tree).
+
+Target releases: 1.15, 1,16, 1.17
+
+Actions:
+
+- Mark in-tree dockershim as "maintenance mode":
+  - CRI generic changes/features can continue on dockershim.
+  - WIP efforts on dockershim can continue and go to complete.
+  - dockershim/docker specific changes/features should be rejected.
+- Revisit and fix dependency issues of dockershim/docker in kubelet (e.g. #54405). Currently, kubelet still has:
+  - vendored dockershim 
+  - flags that are used to configure dockershim.
+  - support to get container logs when docker uses journald as the driver.
+  - logic of moving docker processes to a given cgroup
+  - TBD anything else?
+- Package in-tree dockershim separately from kubelet and provide a "option" to enable/disable it.
+- Ensure e2e/Node e2e test framework is CRI generic and test cases are independent of container runtime.
+
+Step 2: Work out a out-of-tree CRI for docker
+
+Target releases: 1.18
+
+Actions:
+
+- Design & implement a out-of-tree CRI for docker, it can be "copied" from dockershim as beginning.
```

Comment:
I wish we could stop shipping new features (except critical bugfix) to build-in dockershim as early as possible, we could mark it as "deprecated" status and lock to specific version, in this case, new CRI functions should return "Not implemented". 

Otherwise, I'm afraid the deprecating process would be blocked by upcoming features in dockershim.